src/main/java/com/linkedin/kafka/clients/largemessage/LargeMessage.java

```diff
@@ -72,43 +73,45 @@ public long startingOffset() {
 
   public String toString() {
     return String.format("[TopicPartition:%s, UUID:%s, NumberOfSegments:%d, MessageSize:%d, BufferedBytes:%d]",
-        _tp, _messageId, _numberOfSegments, _messageSize, _bufferedBytes);
+        _tp, _messageId, _numberOfSegments, _originalValueSize, _bufferedBytes);
   }
 
   private void validateSegment(LargeMessageSegment segment) {
-    int segmentSize = segment.payload.remaining();
-    int seq = segment.sequenceNumber;
+    int segmentSize = segment.segmentByteBuffer().remaining();
+    int seq = segment.sequenceNumber();
 
     if (segmentSize <= 0) {
       throw new InvalidSegmentException("Invalid segment: " + segment + ". Segment size should be greater than 0.");
     }
 
-    if (_messageSize != segment.messageSizeInBytes
-        || _numberOfSegments != segment.numberOfSegments) {
-      throw new InvalidSegmentException("Detected UUID conflict. Segment: " + segment);
+    if (_originalValueSize != segment.originalValueSize()
+        || _numberOfSegments != segment.numberOfSegments()) {
+      throw new InvalidSegmentException("Segment number of offsets does not equal the known number of offsets for segment: " + segment);
     }
 
-    if (!_segments.containsKey(seq) && _bufferedBytes + segmentSize > _messageSize) {
+    if (!_segments.containsKey(seq) && _bufferedBytes + segmentSize > _originalValueSize) {
       throw new InvalidSegmentException("Invalid segment: " + segment + ". Segments have more bytes than the " +
-          "message has. Message size =" + _messageSize + ", segments total bytes = " +
+          "message has. Message size =" + _originalValueSize + ", segments total bytes = " +
           (_bufferedBytes + segmentSize));
     }
   }
 
   private byte[] assembleMessage() {
-    if (_bufferedBytes != _messageSize) {
+    if (_bufferedBytes != _originalValueSize) {
       throw new InvalidSegmentException("Buffered bytes in the message should equal to message size."
-          + " Buffered bytes = " + _bufferedBytes + "message size = " + _messageSize);
+          + " Buffered bytes = " + _bufferedBytes + "message size = " + _originalValueSize);
     }
-    byte[] serializedMessage = new byte[_messageSize];
+    byte[] serializedMessage = new byte[_originalValueSize];
     int segmentStart = 0;
     for (int i = 0; i < _numberOfSegments; i++) {
       ByteBuffer payload = _segments.get(i);
       int payloadSize = payload.remaining();
       payload.get(serializedMessage, segmentStart, payloadSize);
       segmentStart += payloadSize;
     }
-    assert (segmentStart == _messageSize);
+    if (segmentStart !=  _originalValueSize) {
+      throw new IllegalStateException("segmentStart " + segmentStart + " != originalValueSize " + _originalValueSize);
```

Comment:
end offset is not tracked anywhere in this class or method.  I've added everything else.